src/Lean/Compiler/IR/EmitLLVM.lean

```diff
@@ -0,0 +1,2425 @@
+
+inductive Error where
+| unknownDeclaration: Name → Error
+| invalidExportName: Name → Error
+| unimplemented: String → Error
+| compileError: String → Error -- TODO: these gotta be changed into real errors
+
+instance : ToString Error where
+  toString e := match e with
+   | .unknownDeclaration n => s!"unknown declaration '{n}'"
+   | .invalidExportName n => s!"invalid export name '{n}'"
+   | .unimplemented s => s!"unimplemented '{s}'"
+   | .compileError s => s!"compile error '{s}'"
+
+
+abbrev M := StateT State (ReaderT Context (ExceptT Error IO))
+
+instance : Inhabited (M α) where
+  default := throw (Error.compileError "inhabitant")
+
+
+def addVartoState (x: VarId) (v: LLVM.Ptr LLVM.Value): M Unit :=
+  modify (fun s => { s with var2val := s.var2val.insert x v}) -- add new variable
+
+def addJpTostate (jp: JoinPointId) (bb: LLVM.Ptr LLVM.BasicBlock): M Unit :=
+  modify (fun s => { s with jp2bb := s.jp2bb.insert jp bb })
+
+def emitJp (jp: JoinPointId): M (LLVM.Ptr LLVM.BasicBlock) := do
+  let state ← get
+  match state.jp2bb.find? jp with
+  | .some bb => return bb
+  | .none => throw (Error.compileError s!"unable to find join point {jp}")
+
+/-
+def getEnv : M Environment := Context.env <$> read
+def getModName : M Name := Context.modName <$> read
+def getDecl (n : Name) : M Decl := do
+  let env ← getEnv
+  match findEnvDecl env n with
+  | some d => pure d
+  | none   => throw s!"unknown declaration '{n}'"
+-/
+def getLLVMContext : M (LLVM.Ptr LLVM.Context) := Context.llvmctx <$> read
+def getLLVMModule : M (LLVM.Ptr LLVM.Module) := Context.llvmmodule <$> read
+def getEnv : M Environment := Context.env <$> read
+def getModName : M Name := Context.modName <$> read
+def getDecl (n : Name) : M Decl := do
+  let env ← getEnv
+  match findEnvDecl env n with
+  | some d => pure d
+  | none   => IO.eprintln "getDecl failed!"; throw (Error.unknownDeclaration n)
+
+
+def debugPrint (s: String): M Unit :=
+  IO.eprintln $ "[debug:" ++ s ++ "]"
+
+def constIntUnsigned (n: Nat): M (LLVM.Ptr LLVM.Value) :=  do
+    LLVM.constIntUnsigned (← getLLVMContext) (UInt64.ofNat n)
+
+-- vv emitMainFnIfIneeded vv
+def getOrCreateFunctionPrototype (_ctx: LLVM.Ptr LLVM.Context) (mod: LLVM.Ptr LLVM.Module)
+  (retty: LLVM.Ptr LLVM.LLVMType) (name: String) (args: Array (LLVM.Ptr LLVM.LLVMType)): M (LLVM.Ptr LLVM.Value) := do
+  -- void lean_initialize();
+  LLVM.getOrAddFunction mod name $
+     (← LLVM.functionType retty args (isVarArg := false))
+
+-- ***lean_box***
+def getOrCreateLeanBoxFn: M (LLVM.Ptr LLVM.Value) := do
+  let ctx ← getLLVMContext
+  getOrCreateFunctionPrototype ctx (← getLLVMModule)
+    (← LLVM.voidPtrType ctx) "lean_box"  #[(← LLVM.size_tType ctx)]
+
+def callLeanBox (builder: LLVM.Ptr LLVM.Builder) (arg: LLVM.Ptr LLVM.Value) (name: String := ""): M (LLVM.Ptr LLVM.Value) := do
+  LLVM.buildCall builder (← getOrCreateLeanBoxFn) #[arg] name
+
+-- ***void lean_mark_persistent (void *) ***--
```

Comment:
I think you've upset GH's code highlighter 
![image](https://user-images.githubusercontent.com/109126/202173481-7f6cc385-755b-413f-849e-b97127eb0787.png)